### DISCUSSION TITLE
OpenDTU Logger

### DISCUSSION THREAD

**Discussion by pieterhollander** [tbnobody/OpenDTU#1842] [Show and tell]
![Screenshot 2024-03-18 at 13-03-42 PV2 - Dashboards - Grafana](https://github.com/tbnobody/OpenDTU/assets/1392118/c4591bb5-572a-46a0-8b33-dc4129199f8e)

Hello everyone, 

I wanted to efficiently log long-term production data from my OpenDTU into a database and I've created a program to do just that. Today I'm open sourcing the program. OpenDTU logger is available in binary form for Linux, macOS and Windows, but the preferred way to run it is using Docker Compose. That way, automatic updates can also be configured.

As the OpenDTU software is designed to run on ESP32 boards without a large amount of storage, it does not store historical production data. Instead, users are supposed to rely on individual MQTT topics to which it publishes metrics for this. For me, this was not convenient as it implies storing each metric individually as a “timestamp, metric, value” tuple. I built a proof-of-concept using this approach, but although functional, it was inefficient in terms of storage requirements and data processing.

Instead, I’d like to make use of normalised, SQL-based tables, formatted like “timestamp, metric1, metric2, etc.”. Fortunately, the web interface of OpenDTU relies on a websocket connection, which provides an output more suitable for this, as it outputs, for each inverter, all data at once, in a JSON-based structure.

After not being able to find standalone implementations capable of logging the OpenDTU output into a database efficiently, outside of integrated solutions for platforms such as Home Assistant, I decided to write something myself. I have tested it myself for the past few months and am happy with how it performs.

For more information about the technical design choices I made, [read the blog post I made about it](https://pieterhollander.nl/post/opendtu-logger/). The source code, Docker images and binaries for Windows, Linux and macOS, combined with installation instructions are [available on my personal git server](https://git.hollander.online/energy/opendtu-logger).

Attached is an anonymised screenshot of one of the Grafana dashboards I am able to create with it.

**Comment by pieterhollander** [comment]
It's been a few weeks, and meanwhile I have created version 0.0.6.

OpenDTU logger connects to the OpenDTU livedata websocket and captures metrics. These metrics are inserted into a PostgreSQL database. Optionally, TimescaleDB can be used. Included in OpenDTU Logger are a number of Grafana dashboards, which can be used to provide useful visualisations and insights. The image below shows the PV dashboard of a given day in dark mode. Light mode is also supported. Hovering over the graphs will result in a tooltip, providing exact the values recorded at a given time.
In this release:

[README.md](https://git.hollander.online/energy/opendtu-logger) has been expanded with a better introduction and [detailed installation instructions using Docker Compose](https://git.hollander.online/energy/opendtu-logger#docker-compose). 

A `systemd` example service file has been added to the repository and can be used when, instead of using Docker, you want to install the binary on Linux, [as detailed in the instructions provided in the README](https://git.hollander.online/energy/opendtu-logger#binary).

The previous issue of requiring manual restarts on OpenDTU disconnects has been resolved both in Docker as well as systemd.

Additionally, this release adds support for the following processor architectures:

- riscv64
- arm/v6 (Raspberry Pi 1 / zero)
- arm/v7 (Raspberry Pi 2 and up)
- i386
- mips64 (some routers supporting OpenWRT)

Processor architectures arm64 and amd64 were already supported.

Feedback to this release can be provided and questions can be asked in this topic or through the [contact form](https://pieterhollander.nl/#contact) on my website.

Attached are two more (anonymised) screenshots of the 'PV' dashboard. One in light and another one in dark mode. The dark mode screenshot shows a full day of PV history. In the light mode screenshot, I set the date range to the last 3 hours.
![Screenshot 2024-04-09 at 13-07-58 PV - Dashboards - Grafana](https://github.com/tbnobody/OpenDTU/assets/1392118/56f45baa-28c8-4a28-8f46-5b59283a8507)
![PV-overview-dark-mode](https://github.com/tbnobody/OpenDTU/assets/1392118/46af72d5-630f-4021-904e-d486f62f7ce8)

**Comment by pieterhollander** [comment]
In these screenshots, you can see how hovering over the graphs provides detailed information.

<img width="753" alt="Screenshot 2024-04-09 at 13 29 44" src="https://github.com/tbnobody/OpenDTU/assets/1392118/c728b873-9c32-4ecf-983f-7491ab19699b">
<img width="1488" alt="Screenshot 2024-04-09 at 13 29 36" src="https://github.com/tbnobody/OpenDTU/assets/1392118/17b0548c-808f-4df5-8c9c-81999fe26c13">

**Comment by pieterhollander** [comment]
Today, I'm releasing v0.0.7. In this release, event logging has been implemented to work properly.

Major features

- Event logging was improved to accurately record events and prevent duplication.
- IMPORTANT: In order for event logging to work properly, follow the [instructions on how to configure OpenDTU](https://git.hollander.online/energy/opendtu-logger/src/tag/v0.0.7#configuring-opendtu).

Other improvements

- Miscellaneous edits were made to README.md, better explaining the goals of OpenDTU Logger and providing information on how to give feedback.

**Comment by pieterhollander** [comment]
Hello everyone,

I just finished work on creating a **Home Assistant add-on** for [OpenDTU Logger](https://git.hollander.online/energy/opendtu-logger) and released [version 0.0.9](https://git.hollander.online/energy/opendtu-logger/releases).

OpenDTU Logger allows you to store detailed inverter metrics into a database, which can be visualised using Grafana and other tools.
If you’re interested, please [check it out](https://git.hollander.online/energy/home-assistant-addons/src/branch/main/opendtu-logger) and provide me with your feedback in this topic or through other means.

**Comment by snakekick** [comment]
Really fantastic project, which I am now also using.
Unfortunately, I'm absolutely incompetent when it comes to sql queries. You have created an example grafan dashboard. Perhaps if you have the time and inclination, you could build a query that shows the quantity produced per day for the last 30 days?

Thank you very much for what you have created
Matthias

**Comment by snakekick** [comment]
Hello,
I have created a customised and localised (german) dashboard with the display of the yield of the last 30 days per day

 https://grafana.com/grafana/dashboards/23432-pv/

**Comment by firex2** [comment]
@pieterhollander I tried your solution and dashboard. It works great! Except one thing: in your default dashboard, the "Power per solar panel
" only shows one entry, and not all panels connected to all inverts. How can I proceed debugginf so it works to show all panels on all inverters?

**Reply by pieterhollander** [reply]
@firex2 Thanks, that great to hear! That sounds strange to me, as the logger either has or doesn't have a connection, it shouldn't only partially log. Can you perhaps post one or more screenshots so I may better understand what issue you're facing?

**Reply by firex2** [reply]
<img width="1079" height="1393" alt="Screenshot_20260626-203022" src="https://github.com/user-attachments/assets/27c147c7-ebf5-46c2-b923-c18cf9e1f55b" />
It looks like that , seems to get a lot of values and only creating one entry

**Reply by firex2** [reply]
@pieterhollander does this look familiar for you?

**Comment by bahoop** [comment]
@pieterhollander - i've send you a message using you homepages contact formular regarding how to contribute (if you like), did you received it? I developed a healthprobe cause i'm using Opendtu in docker locally and it crashes sometimes when Proxmox Backup Server freezes the container.

**Reply by pieterhollander** [reply]
Hi @bahoop and thanks for reaching out. Yes, I did receive it
I'm currently on holiday, but replying to you is on my TODO list.
I'll get back to you once I return, probably early September.
